src/Decompiler/Analysis/IndirectCallRewriter.cs

```diff
@@ -0,0 +1,213 @@
+﻿#region License
+/* 
+ * Copyright (C) 1999-2016 Pavel Tomin.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using System;
+using System.Diagnostics;
+using System.Linq;
+using Reko.Core;
+using Reko.Core.Code;
+using Reko.Core.Expressions;
+using Reko.Core.Types;
+using Reko.Typing;
+
+namespace Reko.Analysis
+{
+    /// <summary>
+    /// Try to rewrite indirect call statements to Applications using
+    /// user-defined data (e.g. global variables, parameters of procedures).
+    /// </summary>
+    class IndirectCallRewriter : ExpressionTypeAscender
+    {
+        private SsaState ssa;
+        private Program program;
+        private Procedure proc;
+        private SsaIdentifierTransformer ssaIdTransformer;
+
+        public IndirectCallRewriter(Program program, SsaState ssa) :
+            base(program, new TypeStore(), new TypeFactory())
+        {
+            this.program = program;
+            this.proc = ssa.Procedure;
+            this.ssa = ssa;
+            this.ssaIdTransformer = new SsaIdentifierTransformer(ssa);
+        }
+
+        public void Rewrite()
+        {
+            foreach (Statement stm in proc.Statements)
+            {
+                CallInstruction ci = stm.Instruction as CallInstruction;
+                if (ci != null)
+                {
+                    try
+                    {
+                        RewriteCall(stm, ci);
+                    }
+                    catch (Exception ex)
+                    {
+                        Debug.Print(
+                            "Exception {0}" + Environment.NewLine +
+                            "when rewriting indirect call statement: {1}",
+                            ex, stm);
+                    }
```

Comment:
This should be a logged error - we exceptions to be reported by users if they encounter them. Use the pattern of passing in an `IServiceProvider` interface to the constructor.